test/connection.test.ts

```diff
@@ -68,46 +68,276 @@ describe('Connection', function () {
+        stream.on('data', (chunk: Buffer) => {
+          data.push(chunk)
+        })
+      })
+      const clientStream = this.clientConn.createStream()
+      clientStream.write(Buffer.alloc(300000))
+      await new Promise(setImmediate)
+      await this.clientConn.end()
+      await this.serverConn.end()
+      assert.equal(Buffer.concat(data).length, 300000)
+    })
+
+    it('should complete sending all money from server when end is called on server side of the connection', async function () {
       const moneySpy = sinon.spy()
-      const dataSpy = sinon.spy()
+      let totalMoney: number = 0
       this.clientConn.on('stream', (stream: DataAndMoneyStream) => {
-        stream.on('end', () => {
-          console.log('money stream end')
+        stream.setReceiveMax(12000)
+        stream.on('money', (amount) => {
           moneySpy()
+          totalMoney += +amount
         })
       })
-      this.clientConn.on('stream', (stream: DataAndMoneyStream) => {
-        stream.on('finish', () => {
-          console.log('data strem end')
+      const serverStream = this.serverConn.createStream()
+      serverStream.setSendMax(6000)
+      await this.serverConn.end()
+      assert.equal(totalMoney, 12000)
+      assert.callCount(moneySpy, 6)
+    })
+
+    it('should complete sending all money from client when end is called on client side of the connection', async function () {
+      const moneySpy = sinon.spy()
+      let totalMoney: number = 0
+      this.serverConn.on('stream', (stream: DataAndMoneyStream) => {
+        stream.on('money', (amount) => {
+          moneySpy()
+          totalMoney += +amount
+        })
+      })
+      const clientStream = this.clientConn.createStream()
+      clientStream.setSendMax(12000)
+      await this.clientConn.end()
+      assert.equal(totalMoney, 6000)
+      assert.callCount(moneySpy, 12)
+    })
+
+    it('should keep connection open when a stream is ended', async function () {
+      const stream1 = this.clientConn.createStream()
+      const stream2 = this.clientConn.createStream()
+      const dataSpy = sinon.spy()
+      const moneySpy = sinon.spy()
+      this.serverConn.on('stream', (stream: DataAndMoneyStream) => {
+        stream.on('data', (data: Buffer) => {
           dataSpy()
         })
+        stream.on('money', (amount) => {
+          moneySpy()
+        })
       })
-      this.serverConn.createStream().setSendMax(100)
-      this.serverConn.createStream().write('hello')
 
-      console.log('about to end')
-      await this.clientConn.end()
-      console.log('ended')
+      stream1.write('hello')
+      await new Promise(setImmediate)
+      stream1.setSendMax(100)
+      await new Promise(setImmediate)
+      await stream1.end()
 
-      assert.callCount(moneySpy, 1)
-      assert.callCount(dataSpy, 1)
+      stream2.write('hello')
+      await new Promise(setImmediate)
+      stream2.setSendMax(200)
```

Comment:
I think a more straightforward way of testing what this test describes would be to just add a listener for the connection close and asserting that it wasn't called